### DISCUSSION TITLE
Whenever doing CRUD operation on a certain models, I also want to *ensure* push the results to Snapshot/Outbox model

### DISCUSSION THREAD

**Discussion by danielpaz6** [prisma/prisma#25113] [Q&A]
### Question

Hi,

I want to implement a snapshot / versioning and also an outbox patterns.
In order to do that, whenever I create/edit/delete certain models, for example Users or Books, I also want to save the result in another model with a JSON attribute

So essentially what I want is the following:

```js
const user = prisma.user.create(...}
prisma.snapshots.create({
  data: {
    payload: user
  }
})
```

The thing is, I want to ensure that both will happen, otherwise, none of them would happen, in order to ensure, I can use `$transaction`, but users may be created in many places in the code, I could have wrap it in a function but I prefer to make a middleware or something like that, is it possible to make a middleware that would do $transaction to save all the content of any operation into a snapshots's payload jsonb model?

**Comment by ludralph** [comment]
Hi @danielpaz6 👋 

You can use middleware to ensure that both the creation/editing/deletion of your main models (e.g., Users or Books) and the saving of the result in another model (e.g., Snapshots) happen within a transaction. This ensures atomicity, meaning either both operations succeed, or neither does.

**Reply by danielpaz6** [reply]
Is there an example? :) I couldn't find any documentation about that.

**Reply by ludralph** [reply]
Hi @danielpaz6 

You could create a custom transaction wrapper that encapsulates the `$transaction` functionality. This approach is described in this [GitHub comment](https://github.com/prisma/prisma/issues/12975#issuecomment-1483175095). It involves creating an extended Prisma client that handles transactions and stores them in a request context. For example:
```ts
const extendedPrismaClient = (client: PrismaClient) =>
  client.$extends({
    client: {
      async $transaction<T>(handler: () => Promise<T>) {
        return Prisma.getExtensionContext(client).$transaction(async (tx) => {
          // Set transaction in context
          RequestContextService.setPrismaTransaction(tx);

          try {
            const result = await handler();
            // Here you could add logic to save to snapshots
            return result;
          } finally {
            RequestContextService.cleanPrismaTransaction();
          }
        });
      },
    },
    query: {
      $allModels: {
        $allOperations({ model, operation, args, query }) {
          const tx = RequestContextService.getPrismaTransaction();
          if (tx) {
            return tx[model][operation](args);
          }
          return query(args);
        },
      },
    },
  });
```
This approach allows you to wrap your operations in a transaction and potentially add snapshot logic within the transaction.
However, it's important to note that this solution is not perfect and may not cover all edge cases, especially for nested operations.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!;
